### DISCUSSION TITLE
Clojupyter 0.3.0 released

### DISCUSSION THREAD

**Discussion by klausharbo** [clojupyter/clojupyter#157] [Announcements]
Clojupyter-0.3.0 has been released.  The focus of the release has been to improve the structure of the implementation, with a much larger share of pure functions resulting in improved testability and clearer separation of layers.  The new kernel communicates with its environment exclusively using `core.async` channels with ZeroMQ communication split off in a separate layer and communication with the internally spawned NREPL server based on a combination of interceptors with core logic implemented as pure functions.

The release fixes reported issues #94 (Use tools.deps for dependencies), #103 (Imagemagick need to install Clojupyter), #104 (Clojupyter not installable in non-base Conda environment), and #105 (Can't interrupt the kernel on Windows), closing all open bugs in the project issue list.

The relase also includes support for Jupyter COMM messages (COMM_OPEN, COMM_CLOSE, COMM, COMM_INFO_REQUEST, COMM_INFO_REPLY) which are used in Jupyter to synchronize state between the front-end (Jupyter Notebook and Lab) and language kernel when using interactive widgets.  The support will serve as the basis for exploring the possibility of adding support for widgets to Clojupyter.

To signify the fairly fundamental changes and additions to Clojupyter in this release the version number is bumped to 0.3.0.

The release is available on Anaconda Cloud as the default install, i.e. it can be installed using the command:

```
conda install -c simplect clojupyter
```

Enjoy!

-Klaus.

**Comment by hickst** [comment]
How does one start Jupyter Lab in this Anaconda installation? The instructions given at [Running Jupyter with Clojupyter](https://github.com/clojupyter/clojupyter#running-jupyter-with-clojupyter) do not seem to work for Jupyter Lab and I can find no `jupyter-lab` command in the anaconda installation bin directory (e.g., mine is /usr/local/src/anaconda3/envs/clojup/bin)

**Comment by klausharbo** [comment]
Hi Tom - 

I have standard installations of Anaconda on my machines.  On my Mac the jupyter executable is 

```
21:21 > which jupyter
/Users/klaush/opt/anaconda3/bin/jupyter
```

running 

```
21:21 > jupyter --version
jupyter core     : 4.5.0
jupyter-notebook : 6.0.1
qtconsole        : 4.5.5
ipython          : 7.8.0
ipykernel        : 5.1.2
jupyter client   : 5.3.3
jupyter lab      : 1.1.4
nbconvert        : 5.6.0
ipywidgets       : 7.5.1
nbformat         : 4.4.0
traitlets        : 4.3.3
(base) hume:~
21:21 >
```

On the Debian Linux virtual machine I use to test Clojputer on Linux the executable is

```
(base) parallels@debian-gnu-linux-8:~$ which jupyter
/home/parallels/anaconda3/bin/jupyter
```

running

```
(base) parallels@debian-gnu-linux-8:~$ jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.2
qtconsole        : 4.6.0
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : 1.2.4
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 4.4.0
traitlets        : 4.3.3
(base) parallels@debian-gnu-linux-8:~$
```

On both I can start Jupyter Lab by either

```
jupyter lab
```

or 

```
jupyter-lab
```

I believe the former is simply a frontend to the latter.

I don't whay might cause your Anaconda installation to not work in this way.  Doing

```
conda install jupyter-lab
```

might help, but I'm really a bit out of my league, as your questions is more about Anaconda and Jupyter than it is about Clojupyter, so checking some of the Anaconda or Jupyter forums might be good idea.

HTH

-Klaus.

**Comment by hickst** [comment]
Thanks Klaus...your details helped to see the problem. To install, I did the following:
```
> conda create -n clojup
> conda activate clojup
> conda install -c simplect clojupyter
```
I assumed that this would install everything **but** my Jupyter is different from yours:
```
> jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.2
qtconsole        : not installed
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : not installed
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 4.4.0
traitlets        : 4.3.3
```
**Note** that Jupyter Lab and QTConsole are missing from the installation when using your `simplect` channel.

**Comment by hickst** [comment]
So, I guess my question is: why does the installation from the `simplect` channel not include Jupyter Lab and QTconsole? Did you create this distribution? Do you plan to add them to the distribution?

**Comment by klausharbo** [comment]
Yes, I did create the distribution.  I guess the reason that Clojupyter does explicitly depend on Lab and Console is that it never occurred to me that was needed because the Anaconda installations I have used always included both so I never encountered the problem.  I am not sure why yours doesn't, but knowing that it can work like that is useful.

And, yes, your posting meant that I concluded that Clojupyter should probably explicitly depend on Lab to ensure that it's available after conda installing Clojupyter.  It does explicitly depend on Notebook, which used to be the default/primary client, but these days Lab is clearly presented as primary by the Jupyter team and so - as you point out - it makes sense to add LAb as an explicit dependency.

So: I have made a note to myself that the next release of Clojupyter should explicitly depend on Lab and hence conda would automatically install it when installing Clojupyter.  

A question then is whether Clojupyter should continue to depend explicitly on Notebook?  I think the answer is 'yes' because I suspect many people still use Notebook and will continue to for quite a while.

I am less sure about depending explicitly Console as I suspect essentially all Clojupyter users would prefer a true Clojure REPL to Console and so having it install by default is not necessary. If you need it, installing it as easy as `conda install qtconsole`.  I have no objections to including it, but don't think it will be much used.

**Comment by klausharbo** [comment]
I just released v0.3.1 which includes explicit dependency on Lab and Console in conda installs so that these two applications are installed by default. :-)

**Comment by hickst** [comment]
@klausharbo Thanks Klaus! I updated my existing Conda environment with
`conda install -c simplect clojupyter` and Jupyter Lab seems to be working well now.

**Comment by klausharbo** [comment]
I think the change makes a lot of sense since people *will* expect - as you do - Lab to be present when you have installed Clojupyter using conda.  So the update is a definite improvement.  

However, re-reading the thread I realize I should perhaps have pointed out that the straightforward solution to the problem of a missing component (here: lab and qtconsole) is to install it using conda.

So, for the record and showing how to use conda to manage components, if we create a completely empty conda environment

```
hume:~ > conda create -y --name show-install
Collecting package metadata (current_repodata.json): done
... elided ...
```
activate and verify that it is empty:

```
hume:~ > conda activate show-install
(show-install) hume:~ > conda list
# packages in environment at /Users/klaush/opt/anaconda3/envs/show-install:
#
# Name                    Version                   Build  Channel
(show-install) hume:~ >
```
install Clojupyter v0.3.0 (which doesn't depend explicitly on Lab and Console):

```
(show-install) hume:~ > conda install -c simplect clojupyter=0.3.0
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/klaush/opt/anaconda3/envs/show-install

  added / updated specs:
    - clojupyter=0.3.0
... elided...
```
verify Clojupyter installed version:

```
(show-install) hume:~ > conda list | grep clojupyter
clojupyter                0.3.0                         1    simplect
```

and Jupyter

```
(show-install) hume:~ > jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.3
qtconsole        : not installed
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : not installed
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 5.0.4
traitlets        : 4.3.3
(show-install) hume:~ >
```
So we're now in the situation Tom was in: No Lab or Console - not what we want.  So we install it:

```
(show-install) hume:~ > conda install jupyterlab qtconsole
Collecting package metadata (current_repodata.json): done
...elided...
```
Review Jupyter versions:

```
(show-install) hume:~ > jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.3
qtconsole        : 4.6.0
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : 1.2.5
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 5.0.4
traitlets        : 4.3.3
(show-install) hume:~ >
```

Lab and Console added - we now have what we want.

Let's remove them again and upgrade Clojupyter to see if that works like we want:

```
(show-install) hume:~ > conda remove -y jupyterlab qtconsole
Collecting package metadata (repodata.json): done
...elided...
```
Lab and Console is no longer installed:

```
(show-install) hume:~ > jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.3
qtconsole        : not installed
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : not installed
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 5.0.4
traitlets        : 4.3.3
(show-install) hume:~ >
```

However, let's upgrade Clojupyter to the new version:

```
(show-install) hume:~ > conda upgrade -y -c simplect clojupyter
Collecting package metadata (current_repodata.json): done
...elided...
The following NEW packages will be INSTALLED:
...elided...
  jupyterlab         conda-forge/noarch::jupyterlab-1.2.5-py_0
...elided...
  qtconsole          conda-forge/noarch::qtconsole-4.6.0-py_0

The following packages will be UPDATED:
  clojupyter                                        0.3.0-1 --> 0.3.1-1
...elided...
(show-install) hume:~ >
```
The **digit after the dash** in the clojupyter version numbers above ('0.3.0-1', '0.3.1-1') is the
**conda build number**, so we're upgrading Clojupyter from version 0.3.0 build 1 to version 0.3.1 build 1.

Checking what we have:

```
(show-install) hume:~ > conda list | grep clojupyter
clojupyter                0.3.1                         1    simplect
(show-install) hume:~ > jupyter --version
jupyter core     : 4.6.1
jupyter-notebook : 6.0.3
qtconsole        : 4.6.0
ipython          : 7.11.1
ipykernel        : 5.1.3
jupyter client   : 5.3.4
jupyter lab      : 1.2.5
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 5.0.4
traitlets        : 4.3.3
(show-install) hume:~ >
```

So, when conda installing Clojupyter v0.3.1, conda kindly installs Lab and Console so we don't have to do it manually. :-)

Reactivate `base` environment and clean up
```
(show-install) hume:~ > conda activate base
(base) hume:~ > conda env remove --name show-install

Remove all packages in environment /Users/klaush/opt/anaconda3/envs/show-install:

(base) hume:~ > conda info --envs
# conda environments:
#
base                  *  /Users/klaush/opt/anaconda3

(base) hume:~ >
```

And we're back to the state we started in.

-Klaus.
